src/Sources.jl

```diff
@@ -0,0 +1,124 @@
+export FileSource, GitSource, DirectorySource
+
+abstract type AbstractSource end
+
+struct FileSource <: AbstractSource
+    url::String
+    hash::String
+end
+
+struct GitSource <: AbstractSource
+    url::String
+    hash::String
+end
+
+struct DirectorySource <: AbstractSource
+    path::String
+end
+
+function download_source(source::FileSource; verbose::Bool = false)
+    if isfile(source.url)
+        # Immediately abspath() a src_url so we don't lose track of
+        # sources given to us with a relative path
+        src_path = abspath(source.url)
+
+        # And if this is a locally-sourced tarball, just verify
+        verify(src_path, source.hash; verbose=verbose)
+    else
+        # Otherwise, download and verify
+        src_path = storage_dir("downloads", string(source.hash, "-", basename(source.url)))
```

Comment:
In this line I had `src_hash` instead of `source.hash`.  This wasn't caught by the tests.... does this mean that tests aren't downloading any file as source? :thinking:

Reply:
That's entirely possible.  They might all be local sources.  :/